subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyHandlerProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,147 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve
+
+import org.gradle.integtests.fixtures.AbstractHttpDependencyResolutionTest
+
+/**
+ * Tests covering the use of {@link org.gradle.api.provider.Provider} as an argument in the
+ * {@link org.gradle.api.artifacts.dsl.DependencyHandler} block.
+ */
+class DependencyHandlerProviderIntegrationTest extends AbstractHttpDependencyResolutionTest {
+
+    def "mutating the provider value before it's added resolves to the correct dependency"() {
+        when:
+        buildFile << """
+        configurations { conf }
+        
+        def lazyDep = objects.property(String).convention("org.mockito:mockito-core:1.8")
+        
+        dependencies {
+            conf lazyDep
+        }
+        
+        lazyDep.set("junit:junit:4.12")
+        
+        task checkDeps {
+            doLast {
+                def deps = configurations.conf.incoming.dependencies
+                assert deps.find { it instanceof ExternalDependency && it.group == 'junit' && it.name == 'junit' && it.version == '4.12' }
+                assert !deps.find { it instanceof ExternalDependency && it.group == 'org.mockito' && it.name == 'mockito-core' && it.version == '1.8'  }
+            }
+        }
+        """
+        then:
+        succeeds('checkDeps')
+    }
+
+    def "works correctly with up-to-date checking"() {
+        given:
+        mavenHttpRepo.module("group", "projectA", "1.1").publish()
+        mavenHttpRepo.module("group", "projectA", "1.2").publish()
+
+        when:
+        buildFile << """
+        repositories {
+            maven {
+                url = "${mavenRepo.uri}"
+            }
+        }
+        configurations { conf }
+        
+        dependencies {
+            conf provider { "group:projectA:\${property('project.version')}" }
+        }
+        
+        task retrieve (type: Sync) {
+            into 'build'
+            from configurations.conf
+        }
+        """
+        then:
+        // First run
+        args('-Pproject.version=1.1')
+        succeeds("retrieve").assertTaskExecuted(":retrieve")
+        // Second run, task should be 'UP-TO-DATE'
+        args('-Pproject.version=1.1')
+        succeeds("retrieve").assertTaskSkipped(":retrieve")
+        // Third run, new version, should not be 'UP-TO-DATE'
+        args('-Pproject.version=1.2')
+        succeeds("retrieve").assertTaskExecuted(":retrieve")
+    }
+
+    def "property has no value"() {
+        when:
+        buildFile << """
+        configurations { conf }
+        
+        def emptyDep = objects.property(String)
+        
+        dependencies {
+            conf emptyDep
+        }
+        // Do the resolve
+        configurations.conf.incoming.dependencies
+        """
+        then:
+        fails("help")
+        errorOutput.contains("No value has been specified for this property.")
```

Comment:
:x: should be `failure.assertHasCause`. Does the error tell what property is in cause?

Reply:
No, there's no 'name' on the property so it can't tell you where it came from.